phoenix-core/src/it/java/org/apache/phoenix/end2end/index/ImmutableIndexIT.java

```diff
@@ -259,7 +276,136 @@ private void assertIndexMutations(Connection conn) throws SQLException {
+                PIndexState indexState = TestUtil.getIndexState(conn, fullIndexName);
+                assertEquals(PIndexState.ACTIVE, indexState);
+            }
+        }
+    }
+
+    @Test
+    public void testGlobalImmutableIndexDelete() throws Exception {
+        if (localIndex || transactionProvider != null) {
+            return;
+        }
+        Properties props = PropertiesUtil.deepCopy(TEST_PROPERTIES);
+        String tableName = "TBL_" + generateUniqueName();
+        String indexName = "IND_" + generateUniqueName();
+        String fullTableName = SchemaUtil.getTableName(TestUtil.DEFAULT_SCHEMA_NAME, tableName);
+        String fullIndexName = SchemaUtil.getTableName(TestUtil.DEFAULT_SCHEMA_NAME, indexName);
+        TABLE_NAME = fullTableName;
+        try (Connection conn = DriverManager.getConnection(getUrl(), props);
+                Admin admin = conn.unwrap(PhoenixConnection.class).getQueryServices().getAdmin();) {
+            conn.setAutoCommit(true);
+            int numRows = 2;
+            createAndPopulateTableAndIndexForConsistentIndex(conn, fullTableName, fullIndexName, numRows);
+
+            String dml = "DELETE from " + fullTableName + " WHERE varchar_pk='varchar1'";
+            conn.createStatement().execute(dml);
+            ResultSet rs;
+            rs = conn.createStatement().executeQuery("SELECT /*+ NO_INDEX */ COUNT(*) FROM " + TABLE_NAME);
+            assertTrue(rs.next());
+            assertEquals(numRows - 1, rs.getInt(1));
+            rs = conn.createStatement().executeQuery("SELECT COUNT(*) FROM " + fullIndexName);
+            assertTrue(rs.next());
+            assertEquals(numRows - 1, rs.getInt(1));
+
+            // Disable data table so that delete fails on data table but index table row remains as unverified
+            admin.disableTable(TableName.valueOf(TABLE_NAME));
+            dml = "DELETE from " + fullTableName + " WHERE varchar_pk='varchar2'";
+            boolean isDeleteFailed = false;
+            try {
+                conn.createStatement().execute(dml);
+            } catch (Exception ex) {
+                isDeleteFailed = true;
+            }
+            assertEquals(true, isDeleteFailed);
+            rs = conn.createStatement().executeQuery("SELECT COUNT(*) FROM " + fullIndexName);
+            assertTrue(rs.next());
+            assertEquals(numRows - 1, rs.getInt(1));
+            verifyRowsForEmptyColValue(conn, fullIndexName, IndexRegionObserver.UNVERIFIED_BYTES);
+
+            // Now read from index and see that we don't get any data
+            rs =
+                    conn.createStatement().executeQuery(
+                            "EXPLAIN SELECT long_pk, varchar_pk, long_col1 FROM " + TABLE_NAME + " WHERE varchar_pk='varchar2' AND long_pk=2");
+            String actualExplainPlan = QueryUtil.getExplainPlan(rs);
+            assertExplainPlan(false, actualExplainPlan, fullTableName, fullIndexName);
+            assertFalse(rs.next());
```

Comment:
The test has failed to delete the rows WHERE varchar_pk='varchar2'. What is the purpose of this select with WHERE varchar_pk='varchar2' AND long_pk=2? and why to expect assertFalse(rs.next())?

Reply:
fixing it. The tests makes sure we use index for that select. rs.next should actually test the select and see it is not returning data.